logsapi/route_handlers.go

```diff
@@ -25,8 +25,22 @@ import (
 	"go.uber.org/zap"
 )
 
-func handleLogEventsRequest(logger *zap.SugaredLogger, logsChannel chan LogEvent) func(w http.ResponseWriter, r *http.Request) {
+func handleLogEventsRequest(
+	logger *zap.SugaredLogger,
+	logsChannel chan LogEvent,
+	metadataAvailable <-chan struct{},
+) func(w http.ResponseWriter, r *http.Request) {
 	return func(w http.ResponseWriter, r *http.Request) {
+		select {
+		case <-metadataAvailable:
+		case <-time.After(100 * time.Millisecond):
```

Comment:
Agree with the smell. I thought about this approach but went with pushback at logsapi level. The reason why I chose this approach is because of 2 reasons:
1. In my tests I was losing a bunch of data in case metadata was not available and the buffer just got lost because the function was shutdown, however, my tests were not replicating real-world usage since I sent a few requests and then stopped. Assuming that a lambda function will be invoked with a sustained throughput the chances of this case being a concern will be less. WDYT?
2. The logsapi sends logs in a batch and the channel accepts them one by one. Due to this, it is possible that a single payload of requests has some events which are ingested and some which are not. This will cause some events to be ingested multiple times. I am not sure how our stack handles duplicate events with the same timestamp so would love to discuss more on this.

Reply:
> In my tests I was losing a bunch of data in case metadata was not available and the buffer just got lost because the function was shutdown, however, my tests were not replicating real-world usage since I sent a few requests and then stopped. Assuming that a lambda function will be invoked with a sustained throughput the chances of this case being a concern will be less. WDYT?

Sorry, I don't understand. Why does changing from explicitly blocking to blocking because the channel is full make a difference? In case I was unclear in my description before, here's roughly what I meant:

1. Receive log events here, send them on the channel; nothing is receiving from the channel initially.
2. First request comes in from the agent with metadata. This event causes the extension to start receiving logs from the channel.

> The logsapi sends logs in a batch and the channel accepts them one by one. Due to this, it is possible that a single payload of requests has some events which are ingested and some which are not. This will cause some events to be ingested multiple times. I am not sure how our stack handles duplicate events with the same timestamp so would love to discuss more on this.

Excellent point. We don't do any deduplication. In theory we could set a document ID for logs and silently discard duplicates, but APM Server does not currently support specifying document IDs like this.

I suppose we would need to handle batches atomically to do what I described. Or implement deduplication, but that's not such an easy lift.

Reply:
> Sorry, I don't understand. Why does changing from explicitly blocking to blocking because the channel is full make a difference?

Apologies, I didn't explain well. The issue happens when the execution environment shuts down before it receives the metadata. In case of explicit blocking, there would be nothing in the channel unless metadata is available however, in the other case the channel can have some data and when the execution env shuts down without getting metadata the data in the channel will be lost. In the explicit blocking case the metadata can be processed by another execution env.

> I suppose we would need to handle batches atomically to do what I described.

One approach can be to create a queue that will allow us to reserve a number of slots. I can give it a try. Should we do it in a separate PR? In this PR, if the assumption in the first point sounds good, we can remove the metadata handling in logsapi and fix the atomic handling of the batch in another PR.

Reply:
> however, in the other case the channel can have some data and when the execution env shuts down without getting metadata the data in the channel will be lost

OK, I think I understand now. Because we're using a buffered channel, logs may end up in the channel but could hypothetically never be consumed. If we don't accept the logs into the channel, then the logs might later be re-delivered to another instantiation of the extension.

> In my tests I was losing a bunch of data in case metadata was not available and the buffer just got lost because the function was shutdown, however, my tests were not replicating real-world usage since I sent a few requests and then stopped. Assuming that a lambda function will be invoked with a sustained throughput the chances of this case being a concern will be less. WDYT?

I don't think we should ever assume that there will be sustained throughput. However...

If we never receive metadata before shutdown, wouldn't that imply an instrumentation problem? And in this case we would never do anything with the logs anyway, even if they are sent to the next instantiation of the extension?

Reply:
> If we never receive metadata before shutdown, wouldn't that imply an instrumentation problem?

I am not sure about this. IIUC, agent will send data via HTTP call and if the lambda function is small it might lead to the call not being successful and the function returning. I observed this behaviour while testing.

<details>
  <summary>Lambda function used for testing</summary>
  
```
import json
from elasticapm import capture_serverless

@capture_serverless()
def lambda_handler(event, context):
    print("This is function log", context.aws_request_id)
    return {
        'statusCode': 200,
        'body': json.dumps('Hello from Lambda!'+context.aws_request_id)
    }
```
</details>